### ISSUE TITLE
iOS硬解播放，一半绿屏

### ISSUE THREAD

**Issue by zywang** [Bilibili/ijkplayer#2124]
请帮看下，下面这个视频，iOS端播放一半绿屏，使用硬解播放。这个可能是什么原因呢？
http://ugcvideo-10038986.video.myqcloud.com/mixiongrecorder/212119_1476773262069.mp4

![img_5212](https://cloud.githubusercontent.com/assets/680810/19467688/11cede74-9545-11e6-99c3-d09e10a1dd55.PNG)

**Comment by zywang** [comment]
补充，这个视频在Android平台是好的，播放没有问题

**Comment by fg5823820** [comment]
我也一样

**Comment by zywang** [comment]
我的这个视频是华为的一款机器录出来的一个视频。
资讯了我一个做编解码的同事，他分析可能是因为录制的时候开启了一个 multi slice 属性，IJK没有对这个slice做兼容导致。

**Comment by bbcallen** [comment]
Works for me on iPad mini2 with iOS 8.3

**Comment by fg5823820** [comment]
我妹子用的华为P9直播，每次直播几乎就会出现在iPad观看半绿情况

**Comment by wangjordy** [comment]
我这边也有出现这个问题， 是在每次播放视频的首帧时会出现这个绿屏（屏幕中间上方显示正常，下方显示为绿屏）的现象， 2秒左右的时间，绿屏显示消失

**Comment by jpxiong** [comment]
@fg5823820 是华为 P9 哪一款？手机具体型号是否可以发来看看

**Comment by rrfeng** [comment]
@zywang @wangjordy @fg5823820 @jpxiong 
请问这个问题有解决吗？目前看来是 P9 录制的视频有可能导致 iOS 播放器硬解的时候出现半绿，同一个视频在 MacOS Safari 上播放同样会半边消失。
具体原因有没有查到呢？

**Comment by teddyla** [comment]
@zywang @rrfeng 
我看了你的这个码流. 编码是multi-slice.  一般都是demux的问题.  (或者mux时候把startCode补正确也可以解决. 因为third-party demux可能写的很烂)   建议你dump下送给解码器的bitstream.  是否demux dequeueAccessUnit没做好(每个slice应该是单独的nal, 很多代码写的差的demux, 不会check first_mb_in_slice这个标志位, 所以导致没有做好framing).   正常好的demux code 是不会有这个问题~~

**Comment by stevechyuan** [comment]
@gitgjogh @teddyla 
有些編碼器是不能關掉multi-slice選項的，有沒有一個具體的方向，可以判斷是否為multi-slice，然後可以在封裝碼流的時候，將多個slice拼裝成一個frame送出?

**Comment by teddyla** [comment]
@stevechyuan  nope.  root cause 是mix时候startCode 没有补好.   但如果你的demux 代码完善也没有问题.   h264 spec. slice header里面第一个栏位first_mb_in_slice 如果不是0  就说明这个frame是multi slice.

**Comment by stevechyuan** [comment]
@teddyla 
感謝回覆，iOS的hardware decoder貌似不支援multi-slice這個選項，所以這段是沒法改了。
我不是很明白你說的mix的時候startCode沒有補好的意思，不知道能否再解釋詳細些，或是留下聯絡方式我私下聯絡你，感激不盡

**Comment by teddyla** [comment]
@stevechyuan 
background knowledge:
一个mp4 file生成，是在enc 吐出bitstream之后。然后a/v interleave mux到mp4 file. h264 start code是在mux时候补的 不是enc吐出来就带的... 

所以某些厂家的mux代码有问题。对multi-slice没有做好，这种情况只能播放器做demux时候自己有额外的dequeueAccessUnit(nal)做好也没问题。 有的demux没有这个multi-slice判断 可能会导致这个问题。 android默认是不会有这个问题的。

player solution:
给你一段参考代码，合入到你的项目中就可以解决.
http://androidxref.com/6.0.1_r10/xref/frameworks/av/media/libstagefright/mpeg2ts/ESQueue.cpp#dequeueAccessUnitH264

812            if (foundSlice) {
813                ABitReader br(nalStart + 1, nalSize);
814                unsigned first_mb_in_slice = parseUE(&br);
815
816                if (first_mb_in_slice == 0) {
817                    // This slice starts a new frame.
818
819                    flush = true;
820                }
821            }

**Comment by stevechyuan** [comment]
@teddyla 
感謝回覆，由於我是在做h264相關的編碼，某些android hardware encode默認就會開啟multi-slice，到了iOS上播放，如果用iOS hardware decode就會同樓主的視頻一般，我沒有做mux到mp4 file的動作。
前面我指的startCode是指h264的nalu length。
只是在想，針對h264 stream，有沒有辦法把多個multi-slice，重新封裝回單一個h264 frame，避開multi-slice的問題。不知道這部分你有沒有什麼想法?
我另外再詳讀一下你給的代碼。

**Comment by teddyla** [comment]
@stevechyuan 
如果ios你无法自己写demux:
因为vendor hw dec的hw指令集设计为ave hevc的input必须是nal format.
每个slice是单独的nal. 所以你只要保证你的multi-slice中的slice每一个slice都补到startcode就好. 
=> 所以mux时候补startCode做好也可以解决. 其实你这个问题就是mux的bug. 对multi-slice没有补好start code

**Comment by stevechyuan** [comment]
@teddyla 
感謝回覆，但是我是封裝h264成flv送出，所以沒有打包成mp4，每個multi-slice都有補上flv header，
但是iOS用ijkplayer用硬體解碼觀看會有部分綠屏，而iOS那端我也無法做修改，才想要在編碼後送出前做處理，但不知道這樣子要怎麼補上startcode，卡關幾天了。

**Comment by teddyla** [comment]
@stevechyuan

**Comment by teddyla** [comment]
@stevechyuan

**Comment by teddyla** [comment]
@stevechyuan 如果播放器demux的逻辑你无法改.  h264 bitstream mux 到flv.  这个流程你是自己做的还是用的third-party sdk?.  mux时候做好就行.   

adobe flash spec说明了 可以不用每个tag就是完整的frame
else if AVCPacketType == 1
One or more NALUs (can be individual slices per FLV packets; that is, full frames are not strictly required)

**Comment by teddyla** [comment]
@stevechyuan  因该是你mux之后的multi-slice的packet 只有一个start code.  而正常的multi-slice frame. each slice is individual nal.  你只要保证你的每个slice是单独的nal就不会有问题 。 如果你不明白我的意思 。 你找一个正常显示的multi slice码流的bit stream和你这个异常的对比下就知道了

**Comment by stevechyuan** [comment]
@teddyla 
感謝回覆。
我從encoder得到的一個multi-slice 264 frame的byte array會是像這樣:
00 00 00 01 65 [I-slice1 data] 00 00 01 65 [I-slice2 data] 00 00 01 65 [I-slice3 data]...
1.如果只把此幀當做單一I 幀打包成AVCPacket (只對I-slice1的start code做處理，拿掉h264 start code "00 00 00 01"，填入flv packet header)，iOS播放端下半部會是綠屏，android的可以正常播放。
2.如果把frame裡每個slice parse出來，打包成多個AVCPacket分別送出，一次只會有一塊有畫面，且畫面會一直跳動。
3.如果把frame裡每個slice的00 00 01 65都封裝成flv header，然後append在一起，一次送出，這樣也沒辦法播..
是不是我有誤會你意思的地方？

**Comment by jinyunfeng** [comment]
@stevechyuan 这个问题解决了吗

**Comment by xuandu** [comment]
解决了吗.同样问题.............不想换软解,求方法
